### PR TITLE
TST: Regression testing for fixed issues

### LINKED ISSUE #5764: BUG: Applying function on column of Groupby object with as_index=False does not select column
```
>>> df = pd.DataFrame([[1, 2], [1, 4], [5, 6]], columns=['A', 'B'])
>>> df
   A  B
0  1  2
1  1  4
2  5  6
[3 rows x 2 columns]
```

Selecting a column of the GroupBy object, still returns all columns:

```
>>> g = df.groupby('A', as_index=False)['B']
>>> g.get_group(1)
   A  B
0  1  2
1  1  4
[2 rows x 2 columns]
>>> g = df.groupby('A', as_index=False)
>>> g.get_group(1)
   A  B
0  1  2
1  1  4
[2 rows x 2 columns]
>>> g.get_group(1)['B']
0    2
1    4
Name: B, dtype: int64
```

So an applied function with `apply` is applied on all columns:

```
>>> df.groupby('A', as_index=False)['B'].apply(lambda x: x.cumsum())
   A  B
0  1  2
1  2  6
2  5  6
[3 rows x 2 columns]
```

With `as_index=True` it works as expected:

```
>>> g = df.groupby('A')
>>> g.get_group(1)
   A  B
0  1  2
1  1  4
[2 rows x 2 columns]

>>> g = df.groupby('A')['B']
>>> g.get_group(1)
0    2
1    4
Name: B, dtype: int64

>>> df.groupby('A')['B'].apply(lambda x: x.cumsum())
0    2
1    6
2    6
dtype: int64
```

A more elaborate example where this turned out:

```
>>> s="""L1  L2  L3
... X   1   200
... X   2   100
... Z   1   15
... X   3   200
... Z   2   10
... Y   1   1
... Z   3   20
... Y   2   10
... Y   3   100"""
>>> 
>>> df = pd.read_csv(StringIO(s), sep="\s+")
>>> df.groupby("L1")["L3"].apply(lambda x: x.order().cumsum()/x.sum())
L1   
X   1    0.200000
    0    0.600000
    3    1.000000
Y   5    0.009009
    7    0.099099
    8    1.000000
Z   4    0.222222
    2    0.555556
    6    1.000000
dtype: float64
```

But if I don't want the X, Y, Z in the index:

```
>>> df.groupby("L1", as_index=False)["L3"].apply(lambda x: x.order().cumsum()/x.sum())
```

return an error as `x` is a dataframe.

### LINKED ISSUE #6783: smaller than microsecond timedelta64 Series are not saved correctly with to_csv
```
In [24]: x = Series(np.around(randn(10) * 100).astype('timedelta64[ns]'))

In [25]: x
Out[25]:
0   -00:00:00.000000
1    00:00:00.000000
2   -00:00:00.000000
3   -00:00:00.000000
4    00:00:00.000000
5   -00:00:00.000000
6   -00:00:00.000000
7   -00:00:00.000000
8   -00:00:00.000000
9   -00:00:00.000000
dtype: timedelta64[ns]

In [26]: x.values
Out[26]: array([-259,   31,  -49,  -41,  141,  -14, -111, -117,  -20,  -35], dtype='timedelta64[ns]')

In [27]: buf = StringIO()

In [28]: x.to_csv(buf)

In [29]: print(buf.getvalue())
0,-00:00:00.000000
1,00:00:00.000000
2,-00:00:00.000000
3,-00:00:00.000000
4,00:00:00.000000
5,-00:00:00.000000
6,-00:00:00.000000
7,-00:00:00.000000
8,-00:00:00.000000
9,-00:00:00.000000
```

### LINKED ISSUE #7072: truncation issue with pd.read_csv
hi, thank you very much for the awesome pandas tool.  I would like to report an issue i noticed and seek guidance on remedial steps.  When i store a long integer into a file using pd.to_csv, it stores the data fine - but when i read it back using pd.read_csv, it messes with the last 3 digits.  When i try to save it back again using to_csv (without any edits), the  numbers in resulting CSV file is different from the original CSV file.  I've illustrated the problem below (notice how 4321113141090630389 becomes 4321113141090630400 and 4321583677327450765 becomes 4321583677327450880):

original CSV file created by pd.to_csv:

> > grep -e 321583677327450 -e 321113141090630 orig.piece 
> > orig.piece:1,1;0;0;0;1;1;3844;3844;3844;1;1;1;1;1;1;0;0;1;1;0;0,,,4321583677327450765
> > orig.piece:5,1;0;0;0;1;1;843;843;843;1;1;1;1;1;1;0;0;1;1;0;0,64.0,;,4321113141090630389
> > 
> > > import pandas as pd
> > > import numpy as np
> > > 
> > > orig = pd.read_csv('orig.piece')
> > > orig.dtypes
> > > Unnamed: 0      int64
> > > aa             object
> > > act           float64
> > > ...
> > > ...
> > > s_act         float64
> > > dtype: object
> > > orig['s_act'].head(6)
> > > 0             NaN
> > > 1    4.321584e+18
> > > 2    4.321974e+18
> > > 3    4.321494e+18
> > > 4    4.321283e+18
> > > 5    4.321113e+18
> > > Name: s_act, dtype: float64
> > > 
> > > orig['s_act'].fillna(0).astype(int).head(6)
> > > 0                      0
> > > 1    4321583677327450880
> > > 2    4321973950881710336
> > > 3    4321493786516159488
> > > 4    4321282586859217408
> > > 5    4321113141090630400
> > > 
> > > orig.to_csv('convert.piece')
> > 
> > grep -e 321583677327450 -e 321113141090630 orig.piece convert.piece
> > orig.piece:1,1;0;0;0;1;1;3844;3844;3844;1;1;1;1;1;1;0;0;1;1;0;0,,,4321583677327450765
> > orig.piece:5,1;0;0;0;1;1;843;843;843;1;1;1;1;1;1;0;0;1;1;0;0,64.0,;,4321113141090630389
> > convert.piece:1,1;0;0;0;1;1;3844;3844;3844;1;1;1;1;1;1;0;0;1;1;0;0,,,4.321583677327451e+18
> > convert.piece:5,1;0;0;0;1;1;843;843;843;1;1;1;1;1;1;0;0;1;1;0;0,64.0,;,4.3211131410906304e+18

could you please help me understand why read_csv jumbles the last three digits?  Its not even a rounding issue, the digits are totally different (like 4321583677327450765 becomes 4321583677327450880 above)  Is it because of the scientific notation that comes in the way - how can we disable it and let pandas treat this data as jus object/string or plan integer/float?

### LINKED ISSUE #7820: Index([1,2,.np.nan]).get_indexer([np.nan]) returns wrong value?
**When this is merged, make the below change in Categorical as well!**

Having `np.nan` in an `Index` is not returning the position of NaN but `-1`:

```
In[5]: from pandas.core.index import _ensure_index
In[6]: import numpy as np
In[7]: idx = _ensure_index([1,2,3,4, np.nan])
In[8]: np.nan in idx
Out[8]: True
In[9]: idx.get_indexer([np.nan])
Out[9]: array([-1])
```

I'm not sure if that's a bug or intended. What happens is that this (new) test for Categoricals fails:

```
        # if nan in levels, the proper code should be set!
        cat = pd.Categorical([1,2,3, np.nan], levels=[1,2,3])
        cat.levels = [1,2,3, np.nan]
        cat[1] = np.nan
        exp = np.array([0,3,2,-1])
        self.assert_numpy_array_equal(cat.codes, exp)

Traceback (most recent call last):
  File "C:\data\external\pandas\pandas\tests\test_categorical.py", line 555, in test_set_item_nan
    self.assert_numpy_array_equal(cat.codes, exp)
  File "C:\data\external\pandas\pandas\util\testing.py", line 99, in assert_numpy_array_equal
    raise AssertionError('{0} is not equal to {1}.'.format(np_array, assert_equal))
AssertionError: [ 0 -1  2 -1] is not equal to [ 0  3  2 -1].
```

### LINKED ISSUE #7883: Unable to retrieve same-index dataframe with the last item per group
The original thread is [here](http://stackoverflow.com/questions/25046252/filling-in-with-the-last-entry-of-the-group?noredirect=1#comment38958154_25046252). 

``` python
> df
                   C         D  E
A    B                           
bar  one   -1.350006  0.260339  2
     three -0.236451 -0.056614  0
flux six   -0.515571 -0.155078  1
     three -0.365032  1.055669  2
foo  five   1.145811  1.514191  0
     one   -0.108427 -0.643262  2
     two    0.286897  0.405798  2
     two   -1.452286  1.149264  2
```

The goal is to get a dataframe with the same index with each entry having the last value in the groups associated with `A`, i.e.

```
> df
                   C         D  E
A    B                           
bar  one   -0.236451 -0.056614  2
     three -0.236451 -0.056614  2
flux six   -0.365032  1.055669  2
     three -0.365032  1.055669  2
foo  five  -1.452286  1.149264  2
     one   -1.452286  1.149264  2
     two   -1.452286  1.149264  2
     two   -1.452286  1.149264  2
```

However, none of the following work:
- `df.groupby(level='A').apply(lambda x: x[-1])`
- `df.groupby(level='A').transform(pd.Series.last)`
- `df.groupby(level='A').transform(lambda x: x.iloc[-1])`

### LINKED ISSUE #8720: BUG/ERR: adding column with float label to Int64Index gives ValueError
When adding a column with a float label to a DataFame with integer column index, you get a `ValueError: cannot insert 0.5, already exists` error with an incorrect message:

```
In [15]: df = pd.DataFrame(np.random.randn(8,4))

In [16]: df
Out[16]:
          0         1         2         3
0 -0.482266 -1.239583 -1.177658  1.676881
1  0.314620 -0.373801 -1.535093  0.829534
2  0.887573 -0.054425  0.313028  0.590081
3 -0.665164 -0.887124  0.720014 -0.361315
4 -0.082880 -0.435376 -0.124390 -0.883354
5  2.122313  0.155019  0.762171 -0.401316
6  1.368269  0.469080 -0.274609 -1.799257
7 -0.631479  0.302715  0.008982 -0.356644

In [17]: df[0.5] = np.NaN
...
ValueError: cannot insert 0.5, already exists

In [18]: df.loc[:,0.5] = np.NaN
...
TypeError: the label [0.5] is not a proper indexer for this index type (Int64Index)
```

Of course, when the column labels are converted to floats (`df.columns = pd.Float64Index(df.columns)`), this does work.

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #5764
- [x] closes #7883
- [x] closes #7820
- [x] closes #8720
- [x] closes #6783
- [x] closes #7072
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/tests/indexes/test_numeric.py [hunk 1]
```diff
@@ -736,6 +736,12 @@ def test_get_indexer(self):
         expected = np.array([0, 1, 1, 2, 2, 3, 3, 4, 4, 5], dtype=np.intp)
         tm.assert_numpy_array_equal(indexer, expected)
 
+    def test_get_indexer_nan(self):
+        # GH 7820
+        result = Index([1, 2, np.nan]).get_indexer([np.nan])
+        expected = np.array([2])
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/test_numeric.py]:
```suggestion
        expected = np.array([2], dtype=np.intp)
```

File: pandas/tests/io/formats/test_to_csv.py [hunk 2]
```diff
@@ -563,3 +564,12 @@ def test_to_csv_na_rep_long_string(self, df_new_type):
...
 
         assert expected == result
+
+    def test_to_csv_timedelta_precision(self):
+        # GH 6783
+        s = pd.Series([1, 1]).astype("timedelta64[ns]")
+        buf = io.StringIO()
+        s.to_csv(buf)
+        result = buf.getvalue()
+        expected = ",0\n0,0 days 00:00:00.000000001\n1,0 days 00:00:00.000000001\n"
```
Comment (simonjayhawkins) [inline pandas/tests/io/formats/test_to_csv.py]:
```suggestion
        expected_rows = [
            ",0",
            "0,0 days 00:00:00.000000001",
            "1,0 days 00:00:00.000000001",
        ]
        expected = tm.convert_rows_list_to_csv_str(expected_rows)
```

**Review by jreback — APPROVED** [review]
lgtm. merge on green.

**Comment by TomAugspurger** [comment]
Fixed one of the 32-bit failures. We may have a remaining one that I don't immediately see a fix for.

```
    def test_to_csv_timedelta_precision(self):
        # GH 6783
        s = pd.Series([1, 1]).astype("timedelta64[ns]")
        buf = io.StringIO()
        s.to_csv(buf)
        result = buf.getvalue()
        expected = ",0\n0,0 days 00:00:00.000000001\n1,0 days 00:00:00.000000001\n"
>       assert result == expected
E       AssertionError: assert ',0\r\n0,0 da...000000001\r\n' == ',0\n0,0 days...0.000000001\n'
E         - ,0

E         ?   -
E         + ,0
E         - 0,0 days 00:00:00.000000001

E         ?                            -
E         + 0,0 days 00:00:00.000000001
E         - 1,0 days 00:00:00.000000001
...
E         
E         ...Full output truncated (3 lines hidden), use '-vv' to show
```

**Comment by jbrockmendel** [comment]
at least one of the failures looks like a np.intp thing

**Comment by TomAugspurger** [comment]
Strange, on the 32-bit linux build we have

```
self = <pandas.tests.indexes.test_numeric.TestInt64Index object at 0xd35cbbcc>

    def test_get_indexer_nan(self):
        # GH 7820
        result = Index([1, 2, np.nan]).get_indexer([np.nan])
        expected = np.array([2], dtype="int64")
>       tm.assert_numpy_array_equal(result, expected)
E       AssertionError: numpy array are different
E       
E       Attribute "dtype" are different
E       [left]:  int32
E       [right]: int64
```

Does that make sense?

Still not sure about the to_csv one. We might want to drop that test / issue from the list and merge the rest.

**Comment by simonjayhawkins** [comment]
@TomAugspurger the original 2 Windows failures are fixed with the suggestions above (tested locally)

**Comment by TomAugspurger** [comment]
Sorry, I missed that! Will push that up.

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**